nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -529,4 +530,369 @@ impl DataStore {
+                        identity: IdentityMetadataCreateParams {
+                            name: "myinstance".parse().unwrap(),
+                            description: "It's an instance".into(),
+                        },
+                        ncpus: 2i64.try_into().unwrap(),
+                        memory: ByteCount::from_gibibytes_u32(16),
+                        hostname: "myhostname".try_into().unwrap(),
+                        user_data: Vec::new(),
+                        network_interfaces:
+                            params::InstanceNetworkInterfaceAttachment::None,
+                        external_ips: Vec::new(),
+                        disks: Vec::new(),
+                        ssh_public_keys: None,
+                        start: false,
+                    },
+                ),
+            )
+            .await
+            .expect("instance must be created successfully");
+
+        let (.., authz_instance) = LookupPath::new(&opctx, &datastore)
+            .instance_id(instance_id)
+            .lookup_for(authz::Action::Modify)
+            .await
+            .expect("instance must exist");
+
+        macro_rules! assert_locked {
+            ($id:expr) => {
+                let instance = dbg!(
+                    datastore
+                        .instance_updater_lock(&opctx, &authz_instance, &$id)
+                        .await
+                )
+                .expect(concat!(
+                    "instance must be locked by ",
+                    stringify!($id)
+                ));
+                assert_eq!(
+                    instance.runtime_state.updater_id,
+                    Some($id),
+                    "instance's `updater_id` must be set to {}",
+                    stringify!($id),
+                );
+            };
+        }
+
+        macro_rules! assert_not_locked {
+            ($id:expr) => {
+                let err = dbg!(datastore
+                    .instance_updater_lock(&opctx, &authz_instance, &$id)
+                    .await)
+                    .expect_err("attempting to lock the instance while it is already locked must fail");
+                assert_eq!(
+                    err,
+                    Error::internal_error("instance is already locked by another saga")
+                );
+            };
+        }
+
+        // attempt to lock the instance from saga 1
+        assert_locked!(saga1);
+
+        // now, also attempt to lock the instance from saga 2. this must fail.
+        assert_not_locked!(saga2);
+
+        // unlock the instance from saga 1
+        datastore
+            .instance_updater_unlock(&opctx, &authz_instance, &saga1)
+            .await
+            .expect("instance must be unlocked by saga 1");
+
+        // now, locking the instance from saga 2 should succeed.
+        assert_locked!(saga2);
+
+        // trying to lock the instance again from saga 1 should fail
+        assert_not_locked!(saga1);
```

Comment:
This is great, I appreciate how readable the test is!

Could we add a test for double-locking / double-unlocking by the same saga?